lib/remote/httpmessage.cpp

```diff
@@ -0,0 +1,193 @@
+
+#include "remote/httpmessage.hpp"
+#include "base/io-engine.hpp"
+#include "base/json.hpp"
+#include "remote/httputility.hpp"
+#include "remote/url.hpp"
+#include <boost/beast/http.hpp>
+#include <fstream>
+#include <string>
+
+using namespace icinga;
+
+/**
+ * Adapter class for Boost Beast HTTP messages body to be used with the @c JsonEncoder.
+ *
+ * This class implements the @c nlohmann::detail::output_adapter_protocol<> interface and provides
+ * a way to write JSON data directly into the body of a @c HttpResponse.
+ *
+ * @ingroup base
+ */
+class HttpResponseJsonWriter : public AsyncJsonWriter
+{
+public:
+	explicit HttpResponseJsonWriter(HttpResponse& msg) : m_Message{msg}
+	{
+		m_Message.body().Start();
+#if BOOST_VERSION >= 107000
+		m_Message.body().Buffer().reserve(m_MinPendingBufferSize);
+#endif /* BOOST_VERSION */
+	}
+
+	~HttpResponseJsonWriter() override { m_Message.body().Finish(); }
+
+	void write_character(char c) override { write_characters(&c, 1); }
+
+	void write_characters(const char* s, std::size_t length) override
+	{
+		auto buf = m_Message.body().Buffer().prepare(length);
+		boost::asio::buffer_copy(buf, boost::asio::const_buffer{s, length});
+		m_Message.body().Buffer().commit(length);
+	}
+
+	void MayFlush(boost::asio::yield_context& yield) override
+	{
+		if (m_Message.body().Size() >= m_MinPendingBufferSize) {
+			m_Message.Flush(yield);
+		}
+	}
+
+private:
+	HttpResponse& m_Message;
+	// Minimum size of the pending buffer before we flush the data to the underlying stream
+	static constexpr std::size_t m_MinPendingBufferSize = 4096;
+};
+
+HttpRequest::HttpRequest(Shared<AsioTlsStream>::Ptr stream) : m_Stream(std::move(stream))
+{
+}
+
+void HttpRequest::ParseHeader(boost::beast::flat_buffer& buf, boost::asio::yield_context yc)
+{
+	boost::beast::http::async_read_header(*m_Stream, buf, m_Parser, yc);
+	base() = m_Parser.get().base();
+}
+
+void HttpRequest::ParseBody(boost::beast::flat_buffer& buf, boost::asio::yield_context yc)
+{
+	boost::beast::http::async_read(*m_Stream, buf, m_Parser, yc);
+	body() = std::move(m_Parser.release().body());
+}
+
+ApiUser::Ptr HttpRequest::User() const
+{
+	return m_User;
+}
+
+void HttpRequest::User(const ApiUser::Ptr& user)
+{
+	m_User = user;
+}
+
+Url::Ptr HttpRequest::Url() const
+{
+	return m_Url;
+}
+
+void HttpRequest::DecodeUrl()
+{
+	m_Url = new icinga::Url(std::string(target()));
+}
+
+Dictionary::Ptr HttpRequest::Params() const
+{
+	return m_Params;
+}
+
+void HttpRequest::DecodeParams()
+{
+	if (!m_Url) {
+		DecodeUrl();
+	}
+	m_Params = HttpUtility::FetchRequestParameters(m_Url, body());
+}
+
+HttpResponse::HttpResponse(Shared<AsioTlsStream>::Ptr stream, HttpServerConnection::Ptr server)
+	: m_Server(std::move(server)), m_Stream(std::move(stream))
+{
+}
+
+void HttpResponse::Clear()
+{
+	ASSERT(!m_SerializationStarted);
+	result(boost::beast::http::status::ok);
+	version(11);
+	base().clear();
+	body() = body_type::value_type{};
+}
```

Comment:
Just had another thought about this: do we really have to list everything that needs clearing manually here? `http::response` provides an `operator=`, so would there be anything wrong with just using this (similar to how your implementation already does for the body)?

```suggestion
void HttpResponse::Clear()
{
	ASSERT(!m_SerializationStarted);
	boost::beast::http::response<body_type>::operator=({});
}
```

Passes the unit tests on my machine.